tests/MongoDB.Bson.Tests/Specifications/bson-corpus/BsonCorpusTestRunner.cs

```diff
@@ -0,0 +1,316 @@
+            {
+                writerSettings.GuidRepresentation = GuidRepresentation.Unspecified;
+            }
+#pragma warning restore 618
+            var json = document.ToJson(writerSettings);
+            return json.Replace(" ", "");
+        }
+
+        private void RunDecodeErrorTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "type", "description", "bson");
+
+            var bson = BsonUtils.ParseHexString(test["bson"].AsString);
+
+            var exception = Record.Exception(() =>
+            {
+                using (var stream = new MemoryStream(bson))
+                using (var reader = new BsonBinaryReader(stream))
+                {
+                    while (!reader.IsAtEndOfFile())
+                    {
+                        _ = BsonSerializer.Deserialize<BsonDocument>(reader);
+                    }
+                }
+            });
+
+            exception.Should().NotBeNull();
+        }
+
+        private void RunParseErrorTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "type", "description", "string");
+
+            var json = test["string"].AsString;
+
+            var exception = Record.Exception(() => BsonDocument.Parse(json));
+
+            exception.Should().NotBeNull();
+        }
+
+        private void RunValidTest(BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(
+                test,
+                "type",
+                "description",
+                "canonical_bson",
+                "canonical_extjson",
+                "relaxed_extjson",
+                "degenerate_bson",
+                "degenerate_extjson",
+                "converted_bson",
+                "converted_extjson",
+                "lossy");
+
+            byte[] cB = null;
+            if (test.Contains("canonical_bson"))
+            {
+                cB = BsonUtils.ParseHexString(test["canonical_bson"].AsString);
+            }
+
+            byte[] dB = null;
+            if (test.Contains("degenerate_bson"))
+            {
+                dB = BsonUtils.ParseHexString(test["degenerate_bson"].AsString);
+            }
+
+            string cEJ = null;
+            if (test.Contains("canonical_extjson"))
+            {
+                cEJ = test["canonical_extjson"].AsString.Replace(" ", "");
+                cEJ = UnescapeUnicodeCharacters(cEJ);
+            }
+
+            string dEJ = null;
+            if (test.Contains("degenerate_extjson"))
+            {
+                dEJ = test["degenerate_extjson"].AsString.Replace(" ", "");
+                dEJ = UnescapeUnicodeCharacters(dEJ);
+            }
+
+            string rEJ = null;
+            if (test.Contains("relaxed_extjson"))
+            {
+                rEJ = test["relaxed_extjson"].AsString.Replace(" ", "");
+                rEJ = UnescapeUnicodeCharacters(rEJ);
+            }
+
+            if (cB != null)
```

Comment:
Here and below test scenarios adhere to specification's valid test cases assertions (.NET has a language-native representation, so it's the second option from specification):
https://github.com/mongodb/specifications/blob/master/source/bson-corpus/bson-corpus.rst#testing-validity